File: plugins/plugin-local-ai/test-inference.ts
#!/usr/bin/env bun
/**
 * Integration test for TypeScript Local AI plugin with actual model inference
 */

import { getLlama, LlamaChatSession } from 'node-llama-cpp';
import path from 'path';
import os from 'os';

const MODELS_DIR = path.join(os.homedir(), '.eliza', 'models');
const SMALL_MODEL = 'tinyllama-1.1b-chat-v1.0.Q4_K_M.gguf';
const EMBEDDING_MODEL = 'bge-small-en-v1.5.Q4_K_M.gguf';

async function testTextGeneration(): Promise<void> {
  console.log('\n🧪 Testing TypeScript Text Generation...');
  console.log('   Model:', SMALL_MODEL);
  
  const modelPath = path.join(MODELS_DIR, SMALL_MODEL);
  console.log('   Path:', modelPath);
  
  const llama = await getLlama();
  console.log('   ✓ Llama loaded');
  
  const model = await llama.loadModel({
    modelPath,
    gpuLayers: 33, // Use GPU for inference
  });
  console.log('   ✓ Model loaded');
  
  const context = await model.createContext({ contextSize: 2048 });
  const session = new LlamaChatSession({ contextSequence: context.getSequence() });
  console.log('   ✓ Session created');
  
  const prompt = 'What is 2 + 2? Answer in one word.';
  console.log('   Prompt:', prompt);
  
  const startTime = Date.now();
  const response = await session.prompt(prompt, { maxTokens: 50 });
  const elapsed = Date.now() - startTime;
  
  console.log('   Response:', response.trim());
  console.log('   Time:', elapsed, 'ms');
  console.log('   ✅ Text Generation PASSED\n');
  
  // Cleanup
  await context.dispose();
  await model.dispose();
}

async function testEmbedding(): Promise<void> {
  console.log('\n🧪 Testing TypeScript Embedding Generation...');
  console.log('   Model:', EMBEDDING_MODEL);
  
  const modelPath = path.join(MODELS_DIR, EMBEDDING_MODEL);
  console.log('   Path:', modelPath);
  
  const llama = await getLlama();
  console.log('   ✓ Llama loaded');
  
  const model = await llama.loadModel({
    modelPath,
    gpuLayers: 0, // Embeddings typically run on CPU
  });
  console.log('   ✓ Embedding model loaded');
  
  const embeddingContext = await model.createEmbeddingContext({
    contextSize: 512,
  });
  console.log('   ✓ Embedding context created');
  
  const text = 'Hello, world!';
  console.log('   Text:', text);
  
  const startTime = Date.now();
  const result = await embeddingContext.getEmbeddingFor(text);
  const elapsed = Date.now() - startTime;
  
  const embedding = [...result.vector];
  console.log('   Dimensions:', embedding.length);
  console.log('   First 5 values:', embedding.slice(0, 5).map(v => v.toFixed(4)));
  console.log('   Time:', elapsed, 'ms');
  console.log('   ✅ Embedding Generation PASSED\n');
  
  // Cleanup
  await embeddingContext.dispose();
  await model.dispose();
}

async function main(): Promise<void> {
  console.log('========================================');
  console.log('TypeScript Local AI Integration Test');
  console.log('========================================');
  console.log('CUDA_VISIBLE_DEVICES:', process.env.CUDA_VISIBLE_DEVICES || '(not set)');
  console.log('Models directory:', MODELS_DIR);
  
  try {
    await testTextGeneration();
    await testEmbedding();
    
    console.log('========================================');
    console.log('✅ ALL TYPESCRIPT TESTS PASSED');
    console.log('========================================');
  } catch (error) {
    console.error('❌ Test failed:', error);
    process.exit(1);
  }
}

main();

File: plugins/plugin-local-ai/python/test_inference.py
#!/usr/bin/env python3
import os
import time
from pathlib import Path

try:
    from llama_cpp import Llama

    HAS_LLAMA_CPP = True
except ImportError:
    HAS_LLAMA_CPP = False
    print("⚠️  llama-cpp-python not installed. Install with: pip install llama-cpp-python")

MODELS_DIR = Path.home() / ".eliza" / "models"
SMALL_MODEL = "tinyllama-1.1b-chat-v1.0.Q4_K_M.gguf"
EMBEDDING_MODEL = "bge-small-en-v1.5.Q4_K_M.gguf"


def test_text_generation() -> None:
    print("\n🧪 Testing Python Text Generation...")
    print(f"   Model: {SMALL_MODEL}")

    model_path = MODELS_DIR / SMALL_MODEL
    print(f"   Path: {model_path}")

    if not model_path.exists():
        print(f"   ❌ Model not found at {model_path}")
        return

    n_gpu_layers = -1 if os.environ.get("CUDA_VISIBLE_DEVICES") else 0
    print(f"   GPU layers: {n_gpu_layers}")

    llm = Llama(
        model_path=str(model_path),
        n_gpu_layers=n_gpu_layers,
        n_ctx=2048,
        verbose=False,
    )
    print("   ✓ Model loaded")

    prompt = "What is 2 + 2? Answer in one word."
    print(f"   Prompt: {prompt}")

    start_time = time.time()
    output = llm(
        prompt=prompt,
        max_tokens=50,
        temperature=0.1,
        echo=False,
    )
    elapsed = (time.time() - start_time) * 1000

    response = output["choices"][0]["text"].strip()
    print(f"   Response: {response}")
    print(f"   Time: {elapsed:.0f} ms")
    print("   ✅ Text Generation PASSED\n")


def test_embedding() -> None:
    print("\n🧪 Testing Python Embedding Generation...")
    print(f"   Model: {EMBEDDING_MODEL}")

    model_path = MODELS_DIR / EMBEDDING_MODEL
    print(f"   Path: {model_path}")

    if not model_path.exists():
        print(f"   ❌ Model not found at {model_path}")
        return

    llm = Llama(
        model_path=str(model_path),
        n_gpu_layers=0,
        embedding=True,
        verbose=False,
    )
    print("   ✓ Embedding model loaded")

    text = "Hello, world!"
    print(f"   Text: {text}")

    start_time = time.time()
    embedding = llm.create_embedding(text)["data"][0]["embedding"]
    elapsed = (time.time() - start_time) * 1000

    print(f"   Dimensions: {len(embedding)}")
    print(f"   First 5 values: {[f'{v:.4f}' for v in embedding[:5]]}")
    print(f"   Time: {elapsed:.0f} ms")
    print("   ✅ Embedding Generation PASSED\n")


def main() -> None:
    print("========================================")
    print("Python Local AI Integration Test")
    print("========================================")
    print(f"CUDA_VISIBLE_DEVICES: {os.environ.get('CUDA_VISIBLE_DEVICES', '(not set)')}")
    print(f"Models directory: {MODELS_DIR}")

    if not HAS_LLAMA_CPP:
        print("\n❌ Cannot run tests without llama-cpp-python")
        print("Install with: pip install llama-cpp-python")
        return

    try:
        test_text_generation()
        test_embedding()

        print("========================================")
        print("✅ ALL PYTHON TESTS PASSED")
        print("========================================")
    except Exception as e:
        print(f"❌ Test failed: {e}")
        import traceback

        traceback.print_exc()
        exit(1)


if __name__ == "__main__":
    main()
